core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java

```diff
@@ -443,18 +455,109 @@ private SearchResult parseToSearchResult(@Nonnull final IMAPResponse[] ir) throw
+            }
+            final List<FetchResponse> fetchResponses = new ArrayList<>();
+            final List<Long> modifiedMessageIdList = new ArrayList<>();
+
+            Long highestModSeq = null;
+
+            for (final IMAPResponse sr: ir) {
+                if (sr.isOK() || sr.isNO()) {
+                    sr.skipSpaces();
+                    if (sr.readByte() != (byte) L_BRACKET) {
+                        continue;
                     }
+                    final String responseCode = sr.readAtom();
+                    if (responseCode.equalsIgnoreCase("HIGHESTMODSEQ")) {
+                        highestModSeq = sr.readLong();
+                    } else if (responseCode.equalsIgnoreCase("MODIFIED")) {
+                        final String modifiedSeqStr = sr.readAtom();
+                        String[] modifiedSeqArray = modifiedSeqStr.replace("[", "").split(",");
+                        for (String str: modifiedSeqArray) {
+                            modifiedMessageIdList.add(Long.valueOf(str));
+                        }
+                    }
+                } else if (sr.keyEquals("FETCH")) {
+                    fetchResponses.add(new FetchResponse(sr));
+                }
+            }
+
+            return new StoreResult(highestModSeq, fetchResponses, modifiedMessageIdList);
+        }
+
+        /**
+         * Parses the responses from Store command and UID Store command to a @{code StoreResult} object.
+         *
+         * @param ir the list of responses from UID search command, the input responses array should contain the tagged/final one
+         * @return StoreResult object constructed based on the given IMAPResponse array,
+         * @throws ImapAsyncClientException when tagged response is not OK and NO or given response length is 0
+         */
+        @Nonnull
+        private FetchResult parseToFetchResult(@Nonnull final IMAPResponse[] ir) throws ImapAsyncClientException, IOException, ProtocolException {
+            if (ir.length < 1) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final Response taggedResponse = ir[ir.length - 1];
+            if (!taggedResponse.isOK()) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final List<FetchResponse> fetchResponses = new ArrayList<>();
+
+            Long highestModSeq = null;
+
+            for (final IMAPResponse sr: ir) {
+                if (sr.isOK()) {
+                    sr.skipSpaces();
+                    if (sr.readByte() != (byte) L_BRACKET) {
+                        continue;
+                    }
+                    String s = sr.readAtom();
+                    if (s.equalsIgnoreCase("HIGHESTMODSEQ")) {
+                        highestModSeq = sr.readLong();
+                    }
+                } else {
+                    fetchResponses.add(new FetchResponse(sr));
```

Comment:
Assuming the response to be Fetch will crash the code, check the key before calling FetchResponse